### PR TITLE
node: improve and document Context and Traits interface

### PR DESCRIPTION
**Pull request by silesky**
This PR is meant to improve the DX for typescript users who want to have hints about what kind of context or traits they may want to pass.

Customer issue: https://github.com/segmentio/analytics-node/issues/349#issuecomment-1241527417

```ts
analytics.track({ event: "foo", context: { ...... } })
```
Used this as the source of truth: https://segment.com/docs/connections/spec/common/#context

I did not remove any types.

### CODE REVIEW

File: packages/core/src/events/interfaces.ts [hunk 1]
```diff
@@ -38,71 +38,137 @@ export interface CoreOptions {
...
    */
   referrer?: {
-    btid?: string
-    urid?: string
+    type?: string
+    name?: string
+    url?: string
+    link?: string
+
+    btid?: string // TODO? where did this come from?
```
Comment (silesky) [inline packages/core/src/events/interfaces.ts:166]: have no idea where this came from -- it's not in https://segment.com/docs/connections/spec/common/#context

File: packages/core/src/events/interfaces.ts [hunk 2]
```diff
@@ -38,71 +38,137 @@ export interface CoreOptions {
...
+    urid?: string // TODO? where did this come from?
```
Comment (silesky) [inline packages/core/src/events/interfaces.ts:167]: not documented in https://segment.com/docs/connections/spec/common/#context

File: packages/core/src/events/interfaces.ts [hunk 3]
```diff
@@ -38,71 +38,136 @@ export interface CoreOptions {
...
+    country?: string
+    /**
+     * @example 40.2964197
+     */
+    latitude?: string
+    /**
+     * @example -76.9411617
+     */
+    longitude?: string
+    region?: string
```
Comment (silesky) [inline packages/core/src/events/interfaces.ts:84]: This is listed in the docs, but there is no example of this.

Comment (chrisradek) [reply packages/core/src/events/interfaces.ts:83]: Just my guess, but I'd see `region` as a more global friendly version of `state`. Province, State, whatever else you call the equivalent.

File: packages/core/src/events/interfaces.ts [hunk 4]
```diff
@@ -28,64 +26,131 @@ export type Integrations = {
...
+ */
+export interface CoreExtraContext {
+  /**
+   * This is usually used to flag an .identify() call to just update the trait, rather than "last seen".
+   */
+  active?: boolean
+
+  /**
+   * Current user’s IP address.
+   * (node only, collected automatically in browser).
```
Comment (pooyaj) [inline packages/core/src/events/interfaces.ts:51]: You can still override IP on the browser as well

File: packages/core/src/events/interfaces.ts [hunk 5]
```diff
@@ -28,64 +26,131 @@ export type Integrations = {
...
+   */
+  ip?: string
+
+  /**
+   * 	Locale string for the current user, for example en-US.
+   *  (node only, collected automatically in browser).
```
Comment (pooyaj) [inline packages/core/src/events/interfaces.ts:57]: similarly can override on browser as well

**Review by pooyaj — APPROVED** [review]
Looks great

File: packages/core/src/events/interfaces.ts [hunk 6]
```diff
@@ -179,34 +249,129 @@ interface PlanEvent {
...
+   * Email address of a user/group.
+   */
+  email?: string
+
+  /**
+   * @example female
+   */
+  gender?: string
+
+  crossDomainId?: string // undocumented?
```
Comment (chrisradek) [inline packages/core/src/events/interfaces.ts:374]: I think we can exclude this.

File: packages/core/src/utils/ts-helpers.ts [hunk 7]
```diff
@@ -0,0 +1,6 @@
+/**
+ * Remove Index Signature
+ */
+export type RemoveIndexSignature<T> = {
+  [K in keyof T as {} extends Record<K, 1> ? never : K]: T[K]
```
Comment (chrisradek) [inline packages/core/src/utils/ts-helpers.ts:5]: I do not fully understand how this works (why the `1`?) but that's really neat that it does.

**Review by chrisradek — APPROVED** [review]
Love these sorts of updates!

File: packages/core/src/utils/ts-helpers.ts (hunk 7 above)
Comment (silesky) [reply packages/core/src/utils/ts-helpers.ts:5]: Us both!

**Pull request merged**